Alluxio client accesses workers via IP address as an alternative when hostname cannot be resolved

**Is your feature request related to a problem? Please describe.**
Alluxio client can only get access to an alluxio worker via the according hostname of the worker. However when the mapping relation of hostname and IP address is not specified in the /etc/hosts of the worker machine or a DNS server, an **Unknownhost Exception** will be thrown. The problem occurs when I run a presto cluster on alluxio with the hosts mapping relation not specified in the hosts file. The alluxio client embedded in presto worker can not resolve the hostname returned from the alluxio master, then the query submitted to presto failed.

**Describe the solution you'd like**
Let alluxio client can access workers via IP address as an alternative when the hostname cannot be resolved.

**Describe alternatives you've considered**
To solve the problem, I asked the Administrator to modify the /etc/hosts provisionally, and it works well. But it  is not a perfect way since once new workers are added to the cluster, such operation is required. And deploying a DNS server is hard to implement as well. 

**Urgency**
High

**Additional context**
Such feature can be switched on by a configuration property.

@maobaolong Here is the presto server log. The hostname in format tdw-A-B-C-D represents the IP address: A.B.C.D. For example tdw-1-2-3-4 represents 1.2.3.4

2020-10-28T10:28:09.975+0800	ERROR	remote-task-callback-6940	io.prestosql.execution.StageStateMachine	Stage 20201028_022609_00003_zqwb8.1 failed
io.prestosql.spi.PrestoException: Error reading from alluxio://xx.xx.xx.xx:19998/****/tmp/hive/warehouse/tpcds_100g.db/test_table/000000_0 at position 0
	at io.prestosql.plugin.hive.orc.HdfsOrcDataSource.readInternal(HdfsOrcDataSource.java:75)
	at io.prestosql.orc.AbstractOrcDataSource.readFully(AbstractOrcDataSource.java:96)
	at io.prestosql.orc.AbstractOrcDataSource.readFully(AbstractOrcDataSource.java:87)
	at io.prestosql.orc.CachingOrcDataSource.readCacheAt(CachingOrcDataSource.java:78)
	at io.prestosql.orc.CachingOrcDataSource.readFully(CachingOrcDataSource.java:89)
	at io.prestosql.orc.OrcReader.<init>(OrcReader.java:125)
	at io.prestosql.orc.OrcReader.<init>(OrcReader.java:93)
	at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.createOrcPageSource(OrcPageSourceFactory.java:236)
	at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.createPageSource(OrcPageSourceFactory.java:159)
	at io.prestosql.plugin.hive.HivePageSourceProvider.createHivePageSource(HivePageSourceProvider.java:178)
	at io.prestosql.plugin.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:105)
	at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:57)
	at io.prestosql.split.PageSourceManager.createPageSource(PageSourceManager.java:64)
	at io.prestosql.operator.TableScanOperator.getOutput(TableScanOperator.java:298)
	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
	at io.prestosql.operator.Driver.processFor(Driver.java:276)
	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1076)
	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.prestosql.$gen.Presto_340_3_gadeb631____20201023_025912_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: alluxio.exception.status.UnavailableException: Target Unavailable. GrpcChannelKey{ClientType=DefaultBlockWorkerClient-Stream, ClientHostname=tdw-xx-xx-xx-xx, ServerAddress=GrpcServerAddress{HostName=tdw-A-B-C-D, SocketAddress=tdw-A-B-C-D:29999}, ChannelId=200ff522-04d8-40d4-9048-95c3305524f0}
	at alluxio.grpc.GrpcChannelBuilder.build(GrpcChannelBuilder.java:144)
	at alluxio.client.block.stream.DefaultBlockWorkerClient.<init>(DefaultBlockWorkerClient.java:95)
	at alluxio.client.block.stream.BlockWorkerClient$Factory.create(BlockWorkerClient.java:58)
	at alluxio.client.block.stream.BlockWorkerClientPool.createNewResource(BlockWorkerClientPool.java:72)
	at alluxio.client.block.stream.BlockWorkerClientPool.createNewResource(BlockWorkerClientPool.java:35)
	at alluxio.resource.DynamicResourcePool.acquire(DynamicResourcePool.java:319)
	at alluxio.resource.DynamicResourcePool.acquire(DynamicResourcePool.java:288)
	at alluxio.client.file.FileSystemContext.acquireBlockWorkerClientInternal(FileSystemContext.java:523)
	at alluxio.client.file.FileSystemContext.acquireBlockWorkerClient(FileSystemContext.java:503)
	at alluxio.client.block.stream.GrpcDataReader.<init>(GrpcDataReader.java:83)
	at alluxio.client.block.stream.GrpcDataReader.<init>(GrpcDataReader.java:49)
	at alluxio.client.block.stream.GrpcDataReader$Factory.create(GrpcDataReader.java:203)
	at alluxio.client.block.stream.BlockInStream.positionedRead(BlockInStream.java:287)
	at alluxio.client.file.AlluxioFileInStream.positionedReadInternal(AlluxioFileInStream.java:265)
	at alluxio.client.file.AlluxioFileInStream.positionedRead(AlluxioFileInStream.java:235)
	at alluxio.hadoop.HdfsFileInputStream.read(HdfsFileInputStream.java:128)
	at alluxio.hadoop.HdfsFileInputStream.readFully(HdfsFileInputStream.java:145)
	at org.apache.hadoop.fs.FSDataInputStream.readFully(FSDataInputStream.java:111)
	at io.prestosql.plugin.hive.orc.HdfsOrcDataSource.readInternal(HdfsOrcDataSource.java:62)
	... 24 more
Caused by: alluxio.exception.status.UnavailableException: Unable to resolve host tdw-A-B-C-D
	at alluxio.exception.status.AlluxioStatusException.from(AlluxioStatusException.java:155)
	at alluxio.exception.status.AlluxioStatusException.fromStatusRuntimeException(AlluxioStatusException.java:214)
	at alluxio.exception.status.AlluxioStatusException.fromThrowable(AlluxioStatusException.java:199)
	at alluxio.security.authentication.AuthenticatedChannelClientDriver.waitUntilChannelAuthenticated(AuthenticatedChannelClientDriver.java:187)
	at alluxio.security.authentication.AuthenticatedChannelClientDriver.startAuthenticatedChannel(AuthenticatedChannelClientDriver.java:164)
	at alluxio.security.authentication.ChannelAuthenticator.authenticate(ChannelAuthenticator.java:103)
	at alluxio.grpc.GrpcChannelBuilder.build(GrpcChannelBuilder.java:129)
	... 42 more
Caused by: alluxio.shaded.client.io.grpc.StatusRuntimeException: UNAVAILABLE: Unable to resolve host tdw-A-B-C-D
	at alluxio.shaded.client.io.grpc.Status.asRuntimeException(Status.java:533)
	at alluxio.shaded.client.io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:449)
	at alluxio.shaded.client.io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at alluxio.shaded.client.io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at alluxio.shaded.client.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
	at alluxio.shaded.client.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
	at alluxio.shaded.client.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
	at alluxio.shaded.client.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
	at alluxio.shaded.client.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at alluxio.shaded.client.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	... 3 more
Caused by: java.lang.RuntimeException: java.net.UnknownHostException: tdw-A-B-C-D
	at alluxio.shaded.client.io.grpc.internal.DnsNameResolver.resolveAll(DnsNameResolver.java:436)
	at alluxio.shaded.client.io.grpc.internal.DnsNameResolver$Resolve.resolveInternal(DnsNameResolver.java:272)
	at alluxio.shaded.client.io.grpc.internal.DnsNameResolver$Resolve.run(DnsNameResolver.java:228)
	... 3 more
Caused by: java.net.UnknownHostException: tdw-A-B-C-D
	at java.base/java.net.InetAddress$CachedAddresses.get(InetAddress.java:797)
	at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1505)
	at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1364)
	at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1298)
	at alluxio.shaded.client.io.grpc.internal.DnsNameResolver$JdkAddressResolver.resolveAddress(DnsNameResolver.java:646)
	at alluxio.shaded.client.io.grpc.internal.DnsNameResolver.resolveAll(DnsNameResolver.java:404)
	... 5 more

--- a/core/common/src/main/java/alluxio/conf/PropertyKey.java
+++ b/core/common/src/main/java/alluxio/conf/PropertyKey.java
@@ -439,6 +439,15 @@ public final class PropertyKey implements Comparable<PropertyKey> {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.ALL)
           .build();
+  public static final PropertyKey SERVICE_USE_IP =
+      new Builder(Name.SERVICE_USE_IP)
+          .setDefaultValue("false")
+          .setDescription("If true, when alluxio.<service_name>.hostname and "
+              + "alluxio.<service_name>.bind.host of a service not specified, "
+              + "use IP as the connect host of the service.")
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
+          .setScope(Scope.ALL)
+          .build();
   public static final PropertyKey TEST_MODE =
       new Builder(Name.TEST_MODE)
           .setDefaultValue(false)
@@ -4770,6 +4779,7 @@ public final class PropertyKey implements Comparable<PropertyKey> {
     public static final String NETWORK_HOST_RESOLUTION_TIMEOUT_MS =
         "alluxio.network.host.resolution.timeout";
     public static final String SITE_CONF_DIR = "alluxio.site.conf.dir";
+    public static final String SERVICE_USE_IP = "alluxio.service.use.ip";
     public static final String TEST_MODE = "alluxio.test.mode";
     public static final String TMP_DIRS = "alluxio.tmp.dirs";
     public static final String USER_LOGS_DIR = "alluxio.user.logs.dir";

--- a/core/common/src/main/java/alluxio/conf/PropertyKey.java
+++ b/core/common/src/main/java/alluxio/conf/PropertyKey.java
@@ -439,8 +439,8 @@ public final class PropertyKey implements Comparable<PropertyKey> {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.ALL)
           .build();
-  public static final PropertyKey SERVICE_USE_IP =
-      new Builder(Name.SERVICE_USE_IP)
+  public static final PropertyKey NETWORK_IP_ADDRESS_USED =
+      new Builder(Name.NETWORK_IP_ADDRESS_USED)
           .setDefaultValue("false")
           .setDescription("If true, when alluxio.<service_name>.hostname and "
               + "alluxio.<service_name>.bind.host of a service not specified, "
@@ -4778,8 +4778,8 @@ public final class PropertyKey implements Comparable<PropertyKey> {
         "alluxio.network.connection.shutdown.timeout";
     public static final String NETWORK_HOST_RESOLUTION_TIMEOUT_MS =
         "alluxio.network.host.resolution.timeout";
+    public static final String NETWORK_IP_ADDRESS_USED = "alluxio.network.ip.address.used";
     public static final String SITE_CONF_DIR = "alluxio.site.conf.dir";
-    public static final String SERVICE_USE_IP = "alluxio.service.use.ip";
     public static final String TEST_MODE = "alluxio.test.mode";
     public static final String TMP_DIRS = "alluxio.tmp.dirs";
     public static final String USER_LOGS_DIR = "alluxio.user.logs.dir";

--- a/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java
+++ b/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java
@@ -253,7 +253,7 @@ public final class NetworkAddressUtils {
    * <tr>
    * <th>Specified Hostname</th>
    * <th>Specified Bind Host</th>
-   * <th>Enable Service Using IP</th>
+   * <th>Enable Network IP Address Used</th>
    * <th>Returned Connect Host</th>
    * </tr>
    * </thead>
@@ -309,7 +309,7 @@ public final class NetworkAddressUtils {
         return bindHost;
       }
     }
-    if (conf.getBoolean(PropertyKey.SERVICE_USE_IP)) {
+    if (conf.getBoolean(PropertyKey.NETWORK_IP_ADDRESS_USED)) {
       return getLocalIpAddress((int) conf.getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS));
     }
     return getLocalHostName((int) conf.getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS));

--- a/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java
+++ b/core/common/src/main/java/alluxio/util/network/NetworkAddressUtils.java
@@ -253,6 +253,7 @@ public final class NetworkAddressUtils {
    * <tr>
    * <th>Specified Hostname</th>
    * <th>Specified Bind Host</th>
+   * <th>Enable Service Using IP</th>
    * <th>Returned Connect Host</th>
    * </tr>
    * </thead>
@@ -260,22 +261,32 @@ public final class NetworkAddressUtils {
    * <tr>
    * <td>hostname</td>
    * <td>hostname</td>
+   * <td>true/false</td>
    * <td>hostname</td>
    * </tr>
    * <tr>
    * <td>not defined</td>
    * <td>hostname</td>
+   * <td>true/false</td>
    * <td>hostname</td>
    * </tr>
    * <tr>
    * <td>hostname</td>
    * <td>0.0.0.0 or not defined</td>
+   * <td>true/false</td>
    * <td>hostname</td>
    * </tr>
    * <tr>
    * <td>not defined</td>
    * <td>0.0.0.0 or not defined</td>
-   * <td>localhost</td>
+   * <td>false</td>
+   * <td>local hostname</td>
+   * </tr>
+   * <tr>
+   * <td>not defined</td>
+   * <td>0.0.0.0 or not defined</td>
+   * <td>true</td>
+   * <td>local IP address</td>
    * </tr>
    * </tbody>
    * </table>
@@ -297,6 +308,9 @@ public final class NetworkAddressUtils {
       if (!bindHost.isEmpty() && !bindHost.equals(WILDCARD_ADDRESS)) {
         return bindHost;
       }
+    }
+    if (conf.getBoolean(PropertyKey.SERVICE_USE_IP)) {
+      return getLocalIpAddress((int) conf.getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS));
     }
     return getLocalHostName((int) conf.getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS));
   }

